### PR TITLE
Labrad data support

### PR DESCRIPTION
**Pull request by xlelephant**
This PR aims to address https://github.com/pyplotter/pyplotter/issues/7 by porting the Labrad backend from the repository provided [here](https://github.com/labrad/servers/tree/master/datavault).
Each folder ending with ".dir" is treated as a Labrad "Database". Additionally, I add double-click to navigation to the parent folder if datasets are stored within a child folder. To efficiently handle large datasets, I utilize dv.listDatasets() to check database items quickly without loading dataset information.

loadLabradDataFromRun.py is responsible for parsing parameters and loading data from a dataset. Notably, to expedite the plotting process, I've opted not to employ mp.Queue here. The performance is almost comparable to Labrad's Grapher.exe.)

Default Labrad backend settings have been adjusted to open HDF5 files with the Single-Writer/Multiple-Reader (SWMR) feature enabled for improved efficiency.

### CODE REVIEW

**Comment by edumur** [comment]
Hi,

I currently do not have time to check such amount of new code.
I will be able to check this in ~2 weeks, I think.

Thanks for the work!

**Comment by xlelephant** [comment]
> Hi,
> 
> I currently do not have time to check such amount of new code. I will be able to check this in ~2 weeks, I think.
> 
> Thanks for the work!

No problem, take your time.
Thank you!

File: pyplotter/pyplotter.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 # Init config file, to be done first
 initConfig()
 config = loadConfigCurrent()
+os.environ['HDF5_USE_FILE_LOCKING'] = config['HDF5_USE_FILE_LOCKING']
```
Comment (edumur) [inline pyplotter/pyplotter.py]: Please add a comment here (basically the same one as in the config file will suffice)

File: pyplotter/sources/functions.py [hunk 2]
```diff
@@ -118,6 +113,16 @@ def isBlueForsFolder(folderName : str) -> bool:
     return len(folderName.split('-'))==3 and all([len(i)==2 for i in folderName.split('-')])
 
 
+def isQcodesData(folderName: str):
```
Comment (edumur) [inline pyplotter/sources/functions.py]: `isQcodesData(folderName: str) -> bool:`

And add docstring ;) .

File: pyplotter/sources/functions.py [hunk 3]
```diff
@@ -203,6 +205,8 @@ def getPlotRef(databaseAbsPath: str,
     # If BlueFors log files
     if isBlueForsFolder(currentPath):
         dataPath = currentPath
+    elif isLabradFolder(databaseAbsPath):
```
Comment (edumur) [inline pyplotter/sources/functions.py:214]: add `# If Labrad data folder`

File: pyplotter/sources/functions.py [hunk 4]
```diff
@@ -606,3 +601,78 @@ def getDialogWidthHeight(nbDialog: int) -> Tuple[List[int]]:
         ys[i] = int(i//2*dialogHeight)
 
     return xs, ys, dialogWidths, dialogHeights
+
```
Comment (edumur) [inline pyplotter/sources/functions.py:615]: This  looks like is not related to labrad but to how livePlot dialog windows are handles.
Could you explain why you changed the code here?

File: pyplotter/sources/functions.py [hunk 5]
```diff
@@ -606,3 +601,78 @@ def getDialogWidthHeight(nbDialog: int) -> Tuple[List[int]]:
...
+        else:
+            xs[i] = int(dialogWidth)
+        ys[i] = int(i // 2 * dialogHeight)
+
+    xs = [x + _xs[plotId] for x in xs]
+    ys = [y + _ys[plotId] for y in ys]
+    return xs, ys, dialogWidths, dialogHeights
+
+
+def plotIdGenerator():
```
Comment (edumur) [inline pyplotter/sources/functions.py]: What is the purpose of this function?

File: pyplotter/sources/labradDatavault.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
+import os
+import numpy as np
+import multiprocess as mp
+import copy
+import labrad
+from typing import Tuple, List  # , Dict,  Union, Optional, final
```
Comment (edumur) [inline pyplotter/sources/labradDatavault.py]: Please remove unused comment

File: pyplotter/sources/labradDatavault.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+CXN = labrad.connect()
+# you might want to specify the correct labrad host and password
+
+
+def name2idx(variables, name, default=np.nan):
```
Comment (edumur) [inline pyplotter/sources/labradDatavault.py]: If you could add docstring and typing to your functions?

File: pyplotter/sources/labradDatavault.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+
+def dep_name(dep):
+    if dep["legend"]:
+        return dep["name"] + f"({dep['legend']})"
+    else:
+        return dep["name"]
+
+
+def getNbTotalRun(databaseAbsPath: str) -> None:
```
Comment (edumur) [inline pyplotter/sources/labradDatavault.py]: return int no?

File: pyplotter/sources/labradDatavault.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+    dv = switch_session_path(databaseAbsPath)
+    nbTotalRun = len(dv.dir()[1])
+    return nbTotalRun
+
+
+def getNbTotalRunmp(databaseAbsPath: str, queueNbRun: mp.Queue):
```
Comment (edumur) [inline pyplotter/sources/labradDatavault.py]: return None

File: pyplotter/sources/workers/loadLabradDataBase.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
+from PyQt5 import QtCore
+import multiprocess as mp
```
Comment (edumur) [inline pyplotter/sources/workers/loadLabradDataBase.py]: not used

File: pyplotter/ui/dialogs/dialogLiveplot.py [hunk 11]
```diff
@@ -1,26 +1,41 @@
 from __future__ import annotations
 from PyQt5 import QtWidgets, QtCore
+from idna import check_bidi
```
Comment (edumur) [inline pyplotter/ui/dialogs/dialogLiveplot.py]: not used, please remove

File: pyplotter/ui/tableWidgetDatabase.py [hunk 12]
```diff
@@ -8,18 +8,56 @@
...
-from ..sources.functions import clearTableWidget, getDatabaseNameFromAbsPath
+from ..sources.functions import clearTableWidget, getDatabaseNameFromAbsPath, isLabradFolder, isQcodesData
 from ..ui.dialogs.dialogComment import DialogComment
 from ..ui.menuExportRun import MenuExportRun
 
 # Get the folder path for pictures
 PICTURESPATH = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'pictures')
 
 
+def LoadDataBaseThread(databaseAbsPath, progressBar):
```
Comment (edumur) [inline pyplotter/ui/tableWidgetDatabase.py]: I would prefer not to override the function here but to use the condition later on in the code.
Also `isMongoDBFolder` is undefined.

**Comment by edumur** [comment]
Hi,

First I apology for the delay.
Second, thanks for the big work you did.

As I can't check labrad features locally I will trust you on that.

I have few comments that I gave you directly on code.
Can you please check them?

General comments:
1. Please add labrad in the setup file and in the lib description 
2. Since you're faking the progressBar in the `loadLabradDatabase` file, could you just load everything at once without thread like for csv files?
3. I can't run labrad on my computer even though I will not use it:
 ```bash
plotter
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Scripts\plotter.exe\__main__.py", line 4, in <module>
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\pyplotter.py", line 10, in <module>
    from .ui.mainWindow import MainApp
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\ui\mainWindow.py", line 7, in <module>
    from .mainWindowui_r2c2 import Ui_MainWindow
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\ui\mainWindowui_r2c2.py", line 352, in <module>
    from ..ui.menuBar import MenuBar
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\ui\menuBar.py", line 11, in <module>
    from .dialogs.dialogLiveplot import DialogLiveplot
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\ui\dialogs\dialogLiveplot.py", line 12, in <module>
    from ...sources.labradDatavault import (
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\sources\labradDatavault.py", line 8, in <module>
    CXN = labrad.connect()
          ^^^^^^^^^^^^^^^^
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Lib\site-packages\labrad\__init__.py", line 42, in connect
    cxn = backend.connect(host=host, port=port, name=name, **kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Lib\site-packages\labrad\backend.py", line 86, in connect
    cxn = future.result(timeout=timeout)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Lib\concurrent\futures\_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Lib\concurrent\futures\_base.py", line 401, in __get_result
    raise self._exception
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Lib\site-packages\labrad\concurrent.py", line 84, in wrapped
    result = yield defer.maybeDeferred(func, *args, **kw)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
twisted.internet.error.ConnectionRefusedError: Connection was refused by other side: 10061: No connection could be made because the target machine actively refused it..
```
This is a major issue as it should work for non labrad users.
5. I was then unable to see how and why you change the liveplot dialog and the mainWindow ui.


Thanks!

**Comment by xlelephant** [comment]
Thanks for reviewing! 

The initial goal was to open Labrad data without connecting the server.
However, I was a bit trapped in testing "SWMR" mode of hdf5 files, as it turned out not as stable as we expected.
Instead, I provide two ways in this PR to read Labrad datasets. 
One is "ONLY" for offline and local file reading; 
Another is to use the Labrad server, like a normal Client, which guarantees no harm to the live data.
See `Readme.md`

> 1. Please add labrad in the setup file and in the lib description

Re: I added pylabrad descriptions and also added a requirements.txt file.

> 2. Since you're faking the progressBar in the loadLabradDatabase file, could you just load everything at once without thread like for csv files?

Re: Good suggestion! `loadLabradDatabase` is removed now. The signal flow is much simpler now: 
image

> 3. I can't run labrad on my computer even though I will not use it:
> This is a major issue as it should work for non labrad users.

Re: Non-Labrad users should not be affected now, even if they choose not to install the pylabrad package.

> 4. I was then unable to see how and why you change the liveplot dialog and the mainWindow ui.

Re: I revert to the default layout. You can also have a look at the "Labrad-type" layout by using `mainWindowui_r2c2.py`;)
This layout may not look well on high-dpi screens, I could do more tweaks.

File: README.md [hunk 13]
```diff
@@ -22,6 +22,7 @@ Currently the following packages are required:
 * qcodes>=0.26.0
 * scikit-rf
 * scipy
+* pylabrad (Labrad data support)
```
Comment (edumur) [inline README.md:27]: I am currently  in favor of putting pylabrad as an optional requirement.
Please update the readme accordingly.

Also the setup.cfg would have to be modified following: [](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html)
Adding an optional section
```
[options.extras_require]
```

File: pyplotter/sources/functions.py (hunk 2 above)
Comment (edumur) [reply pyplotter/sources/functions.py]: Still waiting the return type and the docstring.

File: pyplotter/sources/functions.py (hunk 3 above)
Comment (edumur) [reply pyplotter/sources/functions.py:214]: Still there

File: pyplotter/sources/functions.py [hunk 14]
```diff
@@ -247,10 +250,11 @@ def getPlotTitle(databaseAbsPath: str,
         return '{}<br>{} - {}'.format(title, runId, experimentName)
 
 
-
 def getDatabaseNameFromAbsPath(databaseAbsPath: str) -> str:
         return os.path.basename(databaseAbsPath)[:-3]
 
+def getLabradDataVaultNameFromAbsPath(databaseAbsPath: str) -> str:
+        return os.path.basename(databaseAbsPath)[:-4]
```
Comment (edumur) [inline pyplotter/sources/functions.py:263]: Both `getDatabaseNameFromAbsPath` and `getLabradDataVaultNameFromAbsPath` have their return too much indented

File: pyplotter/sources/functions.py (hunk 4 above)
Comment (edumur) [reply pyplotter/sources/functions.py:615]: Still waiting an explanation ^^

File: pyplotter/sources/labradDatavault.py [hunk 15]
```diff
@@ -0,0 +1,479 @@
+import glob
```
Comment (edumur) [inline pyplotter/sources/labradDatavault.py]: not used

File: pyplotter/ui/tableWidgetFolder.py [hunk 16]
```diff
@@ -203,22 +202,36 @@ def itemClicked_(self, b: Union[int, QtCore.QPoint]) -> None:
...
-            if isBlueForsFolder(currentItem):
+            elif isBlueForsFolder(currentItem):
                 self.blueForsClick()
             # If it is a csv or a s2p file
             elif nextPath[-3:].lower() in ['csv', 's2p']:
                 self.csvClick()
             # If it is a npz file
             elif nextPath[-3:].lower() in 'npz':
                 self.npzClick()
+            elif isinstance(b, QtCore.QPoint):
```
Comment (edumur) [inline pyplotter/ui/tableWidgetFolder.py]: Could you explain in which situation you need this?

File: pyplotter/ui/tableWidgetParameter.py [hunk 17]
```diff
@@ -1,15 +1,77 @@
...
+        )
+    return worker
+
+
+def load_by_run_spec(databaseAbsPath, runId):
+    if isLabradFolder(databaseAbsPath):
+        labrad_dataset = LabradDataset(databaseAbsPath, noisy=False)
+        labrad_dataset.loadDataset(runId)
+        return labrad_dataset 
+    # elif isMongoDBFolder(databaseAbsPath):
```
Comment (edumur) [inline pyplotter/ui/tableWidgetParameter.py]: Please remove

**Review by edumur — COMMENTED** [review]
Hi,

Thanks for your work.
The plotter is running locally even though I still didn't run extensive tests.
Please solve the current pending reviews.

Global remarks:

At least locally I have the following error when double clicking on a run:
```bash
Traceback (most recent call last):
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter\pyplotter\ui\tableWidgetParameter.py", line 603, in slotFillTableWidgetParameter
    plotTitle   = getPlotTitle(databaseAbsPath, runId, self.dataset.name)
                                                       ^^^^^^^^^^^^^^^^^
AttributeError: 'function' object has no attribute 'name'
```
Could be related to the change you made in the `tableWidgetParameter` file line 601.

Your two functions are not used in the same way, could you fixe it:
```
isLabradFolder(databaseAbsPath)
isQcodesData(os.path.split(databaseAbsPath)[-1])
```

File: pyplotter/pyplotter.py (hunk 1 above)
Comment (xlelephant) [reply pyplotter/pyplotter.py]: I removed this line.

File: pyplotter/sources/functions.py (hunk 2 above)
Comment (xlelephant) [reply pyplotter/sources/functions.py]: added.

File: pyplotter/sources/functions.py (hunk 3 above)
Comment (xlelephant) [reply pyplotter/sources/functions.py:214]: added.

File: pyplotter/sources/functions.py (hunk 4 above)
Comment (xlelephant) [reply pyplotter/sources/functions.py:615]: Explained in the docstring.

File: pyplotter/sources/functions.py (hunk 5 above)
Comment (xlelephant) [reply pyplotter/sources/functions.py]: also explained.

File: pyplotter/sources/labradDatavault.py (hunk 6 above)
Comment (xlelephant) [reply pyplotter/sources/labradDatavault.py]: removed.

File: pyplotter/sources/labradDatavault.py (hunk 7 above)
Comment (xlelephant) [reply pyplotter/sources/labradDatavault.py]: Added necessary docstring and typing.

File: pyplotter/sources/labradDatavault.py (hunk 8 above)
Comment (xlelephant) [reply pyplotter/sources/labradDatavault.py]: changed to int.

File: pyplotter/sources/labradDatavault.py (hunk 9 above)
Comment (xlelephant) [reply pyplotter/sources/labradDatavault.py]: added return None.

File: pyplotter/sources/workers/loadLabradDataBase.py (hunk 10 above)
Comment (xlelephant) [reply pyplotter/sources/workers/loadLabradDataBase.py]: removed this file.

File: pyplotter/ui/dialogs/dialogLiveplot.py (hunk 11 above)
Comment (xlelephant) [reply pyplotter/ui/dialogs/dialogLiveplot.py]: removed.

File: pyplotter/ui/tableWidgetDatabase.py (hunk 12 above)
Comment (xlelephant) [reply pyplotter/ui/tableWidgetDatabase.py]: Good! I improved these according to your suggestion.

File: README.md (hunk 13 above)
Comment (xlelephant) [reply README.md:27]: I added extras_require filed, but not fully tested the functionality yet.

File: pyplotter/sources/functions.py (hunk 14 above)
Comment (xlelephant) [reply pyplotter/sources/functions.py:263]: I merged both of them in one function, which is indented correctly.

File: pyplotter/sources/labradDatavault.py (hunk 15 above)
Comment (xlelephant) [reply pyplotter/sources/labradDatavault.py]: removed.

File: pyplotter/ui/tableWidgetFolder.py (hunk 16 above)
Comment (xlelephant) [reply pyplotter/ui/tableWidgetFolder.py]: This is redundant, removed.

File: pyplotter/ui/tableWidgetParameter.py (hunk 17 above)
Comment (xlelephant) [reply pyplotter/ui/tableWidgetParameter.py]: removed.

**Comment by xlelephant** [comment]
Thanks for the code review! I refactored the files that you redlighted.
Two Labrad data reading methods were also tested.

File: pyplotter/sources/labradDatavault.py [hunk 18]
```diff
@@ -0,0 +1,525 @@
...
+except:
+    print("Labrad not installed, you will not able to read labrad datasets!")
+
+CXN = None
+USE_LABRAD_DATAVAULT_SERVER = True
+# maximum plot dependents to avoid opening to many 2D windows, if not specificed by the user
+MAX_NUM_PLOT_DEPS = 8
+
+
+def labrad_urldecode(data_url) -> any:
```
Comment (edumur) [inline pyplotter/sources/labradDatavault.py]: The correct typing is `Any` starting with an uppercase.

File: setup.cfg [hunk 19]
```diff
@@ -38,6 +38,10 @@ install_requires =
     scipy
 
 
+[options.extras_require]
+larbad = pylabrad
```
Comment (edumur) [inline setup.cfg]: Typo
`labrad = pylabrad`

**Review by edumur — COMMENTED** [review]
Thanks again for all your work.

Some remarks:
1. the doubleClick on a run of qcodes database is not working. I just disable the widget row and no plot is displayed,
2. By default the plots hide the interaction panel, please leave it on by default. You can add an option in the menu bar, maybe in preference? About it if you wish.


Would you accept to be "in charge" of labrad thing? I have no idea how labrad works...

**Comment by xlelephant** [comment]
> Thanks again for all your work.
> 
> Some remarks:
> 
> 1. the doubleClick on a run of qcodes database is not working. I just disable the widget row and no plot is displayed,

DoubleClick on a Labrad dataset for quick plotting is wanted by most Labrad users, as they can easily check the data. 
But I did not change the default behavior of that on qcodes database, do you want me also to add this feature to qcodes, OR add a statement no the Tip section of the README?

> 2. By default the plots hide the interaction panel, please leave it on by default. You can add an option in the menu bar, maybe in preference? About it if you wish.

I revert that to show the interaction panel by default. Adding a dynamic option in the menu bar is sweet, but I opt not to add too many buttons on the GUI. Instead, I can add explanations for various settings in README.

> 
> Would you accept to be "in charge" of labrad thing? I have no idea how labrad works...

Sure, I would like to do that, and perhaps future development on general database format based on HDF5/NetCDF files.

**Comment by edumur** [comment]
Arf sorry it was a typo... What I meant was that the doubleClick on a run of qcodes database is not working. When I adoubleClik on a run, it just disable the widget row and no plot is displayed,

**Comment by xlelephant** [comment]
> Arf sorry it was a typo... What I meant was that the doubleClick on a run of qcodes database is not working. When I adoubleClik on a run, it just disable the widget row and no plot is displayed,

Oh sorry, I deleted this by mistake. 
Fixed now, please check, thanks!

**Comment by edumur** [comment]
Hi,

Yep, doubleClick is now working.

I saw you added a config parameter `progressBarMaxNb`.
What happen when an user launched a 4th progressBar which then is not created?
How does the program handles the update and deletion of the non existing progressBar?

**Comment by edumur** [comment]
Can you add yourself on the authors section  of the readme file as the Labrad guy?
Thanks!

**Comment by xlelephant** [comment]
> Hi,
> 
> Yep, doubleClick is now working.
> 
> I saw you added a config parameter `progressBarMaxNb`. What happen when an user launched a 4th progressBar which then is not created? How does the program handles the update and deletion of the non existing progressBar?

If too many dependents are toggled at once(to plot), the layout will get too wide to fill in all these progressBars, so I added this restriction.
Except for being added as the "addPermanentWidget", the 4th progressBar will be updated and poped normally.
Note that the `removeWidget` function completes silently even if the input widget was not added.

**Comment by xlelephant** [comment]
> Can you add yourself on the authors section of the readme file as the Labrad guy? Thanks!

Added, thanks!

**Pull request merged**